README.md
## Laravel Queue for AWS Batch

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Code Coverage][ico-coverage]][link-coverage]

### Supported Versions
| Laravel Version | Package Tag | Supported |
|-----------------|-------------|-----------|
| 6.0.x | 3.0.x | yes |

### Installation
See the table above for package version information, and change the version below accordingly.

Add custom repository to your composer json:
```
    "require": {
        "dnxlabs/laravel-queue-aws-batch": "~3.0.0"
    },
    "repositories": [
        {
            "type": "vcs",
            "url":  "git@github.com:dnxlabs/laravel-queue-aws-batch.git"
        }
    ]
```

Using `composer`, run:

    composer require dnxlabs/laravel-queue-aws-batch ~3.0


### Usage
1. Your Laravel application will need to be dockerized and pushed into a container registry of your choice. The `ENTRYPOINT`
   should be set to `artisan`. 

2. Add a new queue to your `config/queues.php` config file's `connections` array:
```
    [
        'batch' => [
            'driver' => 'batch',
            'table' => 'jobs',
            'queue' => 'first-run-job-queue',
            'jobDefinition' => 'my-job-definition',
            'expire' => 60,
            'region' => 'us-east-1'
        ]
    ]
```
This queue transport depends on being able to write it's queue jobs to a database queue. In this example, it writes it's
jobs to the `jobs` table. You'll need to use the `artisan queue:table` to create a migration to create this table.

3. Create an AWS Batch job queue with the same name as the `queue` config setting. This is where the Batch connector
will push your jobs into Batch. In this case, my queue name would be `first-run-job-queue`.

4. Create a AWS Batch job definition for each queue you define that looks something like this:
```json
{
    "jobDefinitionName": "my-laravel-application",
    "type": "container",
    "parameters": {},
    "retryStrategy": {
        "attempts": 10
    },
    "containerProperties": {
        "image": "<your docker image>",
        "vcpus": 1,
        "memory": 256,
        "command": [
            "queue:work-batch",
            "Ref::jobId",
            "--tries=3"
        ],
        "volumes": [],
        "environment": [],
        "mountPoints": [],
        "ulimits": []
    }
}
```
Here, you configure your container to start, run the `queue:work-batch` command (assuming `artisan` is your entrypoint)
and pass in the name of the queue, `first-run-job-queue` as well as the `Ref::jobId` param, which is passed in when
the Batch connector creates the job.

It is important that you configure a retryStrategy with more "attempts" than you are running `tries` if you provide that
argument. Otherwise, Batch will not retry your job if it fails. Laravel 5.1 does not write to the failed job queue until
the _next_ run after tries has been exceeded by jobs failing. Newer versions will write to the queue in the same run, so
this requirement can be relaxed later.

6. Add the Service Provider to your application:
    * In `config/app.php` add to the `providers` array: `DNXLabs\LaravelQueueAwsBatch\BatchQueueServiceProvider::class`
    
    
### Limitations

#### Delayed Jobs
AWS Batch has no method to delay a job and as it's our runner, we don't have an easy work around. If you require delayed
jobs for your use case, at this point my recommendation would be to use a regular DB queue, and to fire a job into it
which will fire your batch job at the correct time.

[ico-version]: https://img.shields.io/packagist/v/dnxlabs/laravel-queue-aws-batch.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/dnxlabs/laravel-queue-aws-batch/master.svg?style=flat-square
[ico-coverage]: https://img.shields.io/scrutinizer/coverage/g/dnxlabs/laravel-queue-aws-batch/master.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/dnxlabs/laravel-queue-aws-batch
[link-travis]: https://travis-ci.org/dnxlabs/laravel-queue-aws-batch
[link-coverage]: https://scrutinizer-ci.com/g/dnxlabs/laravel-queue-aws-batch/?branch=master


CHANGELOG.md
# Release Notes for 1.x

## [v1.2.0] (2021-03-24)

### Added
* Added the default Handler when creating Worker so that exceptions may be reported during Worker::process

## [v1.1.1] (2018-08-21)

### Fixed
* Fix non-overridden jobs by not setting the containerOverrides key when not provided

## [v1.1.0] (2018-08-20)

### Added
* Added support for containerOverrides ([#11c0184](https://github.com/lukewaite/laravel-queue-aws-batch/pull/30))

## [v1.0.2] (2018-07-20)

### Fixed
Fix: Jobs sometimes run until max tries hit without hitting laravel's queue failure method ([#3a25d14e](https://github.com/lukewaite/laravel-queue-aws-batch/commit/3a25d14e7cb3fb6c8d769d92ce0b93d08961ed3d))

## [v1.0.1] (2018-07-15)

### Fixed
* Correctly set the queue on construction of the BatchJob - fixes an issues where
failing jobs were written into the failed jobs table with the connection name as
the queue name. ([#2aa745b](https://github.com/lukewaite/laravel-queue-aws-batch/pull/29))

## [v1.0.0] (2017-04-09)

### Added
* Support for Laravel 5.2, and 5.3 ([#25](https://github.com/lukewaite/laravel-queue-aws-batch/pull/25))

### Fixed
* Fix: Re-throw exception in command handler, don't explicitly `exit()` ([#5cc05a8](https://github.com/lukewaite/laravel-queue-aws-batch/commit/5cc05a88c497ade72b81916a16384bdb69107bd5))

## [v0.2.1] (2017-04-07)

### Fixed
* Correct sanitation of job names which previously blocked creation of class based batch jobs [#22](https://github.com/lukewaite/laravel-queue-aws-batch/pull/22)
* Fix: Failing jobs throw exception when logging about failing (Array to String Conversion) [#d1a5023](https://github.com/lukewaite/laravel-queue-aws-batch/commit/6118f5bdf18935ce346d9628dcd1670f98d8e238)
* Fix: `BatchQueue` methods `->push` and `->update` now return the `id` of the entry [#d1a5023](https://github.com/lukewaite/laravel-queue-aws-batch/commit/6118f5bdf18935ce346d9628dcd1670f98d8e238)
* Fix: BatchQueue now properly releases failed jobs back into the queue [#d1a5023](https://github.com/lukewaite/laravel-queue-aws-batch/commit/6118f5bdf18935ce346d9628dcd1670f98d8e238)

## v0.2.0 (2017-04-07)

### Added
* Added exceptions on further unsupported operations. These two methods would both silently discard the `$delay` previously and set to 0.
  * `BatchQueue::release()` [#20](https://github.com/lukewaite/laravel-queue-aws-batch/pull/20)
  * `BatchJob::release()` [Diff](https://github.com/lukewaite/laravel-queue-aws-batch/pull/19/files#diff-fb4479932d3da5ac0014681d4beba72cR38)
* More complete test coverage ([#18](https://github.com/lukewaite/laravel-queue-aws-batch/pull/18), [#19](https://github.com/lukewaite/laravel-queue-aws-batch/pull/19))

[Unreleased]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v1.1.1...HEAD
[v1.1.1]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v1.0.2...v1.1.0
[v1.0.2]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v1.0.1...v1.0.2
[v1.0.1]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v0.2.1...v1.0.0
[v0.2.1]: https://github.com/lukewaite/laravel-queue-aws-batch/compare/v0.2.0...v0.2.1


src/Connectors/BatchConnector.php
<?php
/**
 * Laravel Queue for AWS Batch.
 *
 * @author    Luke Waite <lwaite@gmail.com>
 * @copyright 2017 Luke Waite
 * @license   http://www.opensource.org/licenses/mit-license.php MIT
 *
 * @link      https://github.com/dnxlabs/laravel-queue-aws-batch
 */

namespace DNXLabs\LaravelQueueAwsBatch\Connectors;

use Aws\Batch\BatchClient;
use Illuminate\Queue\Connectors\DatabaseConnector;
use Illuminate\Support\Arr;
use DNXLabs\LaravelQueueAwsBatch\Queues\BatchQueue;

class BatchConnector extends DatabaseConnector
{
    /**
     * Establish a queue connection.
     *
     * @param array $config
     *
     * @return \Illuminate\Contracts\Queue\Queue
     */
    public function connect(array $config)
    {
        return new BatchQueue(
            $this->connections->connection(Arr::get($config, 'connection')),
            $config['table'],
            $config['queue'],
            Arr::get($config, 'expire', 60),
            $config['jobDefinition'],
            new BatchClient([
                'region'  => $config['region'],
                'version' => '2016-08-10',
            ])
        );
    }
}
